
# Проверка гипотез с помощью python.

## Задание

* Выполненить проверку предложенной гипотезы
* Выполненить проверку статистической гипотезы для проверки нормальности числовых признаков
* Проверка мультиколлинеарности
* Все статистические тесты должны быть выполнены с 95%-ным уровнем достоверности (т.е. значение р < 0,05)

## Датасет

Как данные представлены заказчиком:

<table>
<thead><tr>
<th><strong>Feature Name</strong></th>
<th><strong>Description</strong></th>
<th><strong>Data Type</strong></th>
</tr>
</thead>
<tbody>
<tr>
<td>customerID</td>
<td>Содержит идентификатор клиента</td>
<td>categorical</td>
</tr>
<tr>
<td>gender</td>
<td>Пол клиента</td>
<td>categorical</td>
</tr>
<tr>
<td>SeniorCitizen</td>
<td>Является ли клиент пожилым гражданином</td>
<td>numeric, int</td>
</tr>
<tr>
<td>Partner</td>
<td>Есть ли у клиента партнер</td>
<td>categorical</td>
</tr>
<tr>
<td>Dependents</td>
<td>Является ли клиент кормильцем</td>
<td>categorical</td>
</tr>
<tr>
<td>tenure</td>
<td>Количество месяцев, в течение которых клиент оставался в компании</td>
<td>numeric, int</td>
</tr>
<tr>
<td>PhoneService</td>
<td>Есть ли у клиента телефонная связь</td>
<td>categorical</td>
</tr>
<tr>
<td>MultipleLines</td>
<td>Есть ли у клиента несколько линий</td>
<td>categorical</td>
</tr>
<tr>
<td>InternetService</td>
<td>Интернет-провайдер клиента</td>
<td>categorical</td>
</tr>
<tr>
<td>OnlineSecurity</td>
<td>Есть ли у клиента онлайн-безопасность</td>
<td>categorical</td>
</tr>
<tr>
<td>OnlineBackup</td>
<td>Есть ли у клиента онлайн-резервное копирование</td>
<td>categorical</td>
</tr>
<tr>
<td>DeviceProtection</td>
<td>Имеет ли клиент защиту устройства</td>
<td>categorical</td>
</tr>
<tr>
<td>TechSupport</td>
<td>Есть ли у клиента техническая поддержка</td>
<td>categorical</td>
</tr>
<tr>
<td>streamingTV</td>
<td>Есть ли у клиента потоковое телевидение</td>
<td>categorical</td>
</tr>
<tr>
<td>streamingMovies</td>
<td>Есть ли у клиента стриминговые сервисы</td>
<td>categorical</td>
</tr>
<tr>
<td>Contract</td>
<td>Срок действия контракта заказчика</td>
<td>categorical</td>
</tr>
<tr>
<td>PaperlessBilling</td>
<td>имеет ли клиент безналичные счета</td>
<td>categorical</td>
</tr>
<tr>
<td>PaymentMethod</td>
<td>Способ оплаты клиента</td>
<td>categorical</td>
</tr>
<tr>
<td>MonthlyCharges</td>
<td>Сумма, взимаемая с клиента ежемесячно </td>
<td> numeric , int</td>
</tr>
<tr>
<td>TotalCharges</td>
<td>Общая сумма, списанная с клиента</td>
<td>object</td>
</tr>
<tr>
<td>Churn</td>
<td>Ушел ли клиент</td>
<td>categorical</td>
</tr>
</tbody>
</table>

Описание от заказчика:

Каждая строка представляет клиента, каждый столбец содержит атрибуты клиента.

Набор данных включает информацию о:

- Клиенты, которые ушли в течение последнего месяца - колонка называется `Churn`
- Услуги, на которые подписался каждый клиент - телефон, несколько линий, интернет, онлайн-безопасность, резервное копирование, защита устройств, техническая поддержка, потоковое ТВ и фильмы.
- Информация о счетах клиентов - как долго они являются клиентами, контракт, способ оплаты, безналичные счета, ежемесячные платежи и общая сумма платежей.
- Демографическая информация о клиентах - пол, возраст, наличие партнеров и иждивенцев.

## 1. Импорт Библиотек 

Импортируйте необходимые библиотеки

In [2]:
!pip install scipy pandas numpy


[notice] A new release of pip is available: 24.0 -> 24.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [1]:
import numpy as np
import pandas as pd
pd.set_option('display.max_columns', None) # отобразить все колонки
from scipy import stats

## 2. Данные

In [3]:
df = pd.read_csv("churn.csv")
df.sample(20)

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
4550  4625-XMOYM  Female              0     Yes        Yes      72   
2650  5609-IMCGG  Female              0      No         No      15   
5107  9633-DENPU  Female              0     Yes         No      47   
5894  7881-INRLC    Male              0      No         No      45   
1628  6267-DCFFZ  Female              1     Yes         No      30   
116   9560-BBZXK  Female              0      No         No      36   
3830  2656-TABEH    Male              0     Yes         No       4   
4249  7919-ZODZZ  Female              0     Yes        Yes      10   
4066  4872-VXRIL    Male              0      No         No      56   
6699  3662-FXJFO  Female              0      No         No      54   
2549  0188-GWFLE    Male              0      No         No       2   
6977  2843-CQMEG    Male              0      No         No      24   
314   7563-BIUPC    Male              0      No         No       5   
3396  5578-NKCXI  Female              0     Yes        Yes      58   
5757  5539-HIVAK  Female              1     Yes         No      28   
2791  1023-BQXZE    Male              0      No         No      57   
5534  7964-VEXDG    Male              0      No        Yes      21   
469   4640-UHDOS  Female              0     Yes        Yes      13   
2230  5242-UOWHD    Male              0     Yes        Yes      45   
1298  5989-OMNJE  Female              0     Yes        Yes      24   

     PhoneService     MultipleLines InternetService       OnlineSecurity  \
4550          Yes               Yes              No  No internet service   
2650          Yes               Yes     Fiber optic                  Yes   
5107          Yes                No             DSL                   No   
5894          Yes               Yes     Fiber optic                   No   
1628          Yes               Yes     Fiber optic                   No   
116            No  No phone service             DSL                  Yes   
3830          Yes               Yes     Fiber optic                   No   
4249          Yes                No             DSL                   No   
4066          Yes               Yes             DSL                  Yes   
6699          Yes                No              No  No internet service   
2549          Yes                No              No  No internet service   
6977          Yes                No             DSL                   No   
314           Yes                No     Fiber optic                   No   
3396          Yes                No     Fiber optic                  Yes   
5757          Yes               Yes              No  No internet service   
2791          Yes                No             DSL                  Yes   
5534          Yes               Yes     Fiber optic                  Yes   
469           Yes                No             DSL                   No   
2230          Yes                No              No  No internet service   
1298           No  No phone service             DSL                  Yes   

             OnlineBackup     DeviceProtection          TechSupport  \
4550  No internet service  No internet service  No internet service   
2650                   No                  Yes                   No   
5107                  Yes                   No                   No   
5894                  Yes                   No                   No   
1628                  Yes                   No                   No   
116                    No                   No                   No   
3830                  Yes                   No                   No   
4249                  Yes                  Yes                   No   
4066                  Yes                   No                  Yes   
6699  No internet service  No internet service  No internet service   
2549  No internet service  No internet service  No internet service   
6977                  Yes                   No                   No   
314            

## 2. Описательный анализ данных

Рассмотрите данные - изучите медианы, средние, дисперсии и т.д. 

Что можете сказать о данных? Сделайте вывод.

In [4]:
des=df.describe()
des


SeniorCitizen       tenure  MonthlyCharges
count    7043.000000  7043.000000     7043.000000
mean        0.162147    32.371149       64.761692
std         0.368612    24.559481       30.090047
min         0.000000     0.000000       18.250000
25%         0.000000     9.000000       35.500000
50%         0.000000    29.000000       70.350000
75%         0.000000    55.000000       89.850000
max         1.000000    72.000000      118.750000

**Вывод:**

мы имеем датасет из которого имеем что основная часть клиентов не пожилые люди и достаточно долго остаются в компании 2 - 3 года. Плата за услуги примерно 70

### Уникальные значения для всех столбцов

Рассмотрите уникальные значения категориальных признаков. 

Сделайте вывод.

**Подсказка:** можете воспользоваться методом `describe(include='object')`

In [ ]:
df.describe(include='object')

customerID gender Partner Dependents PhoneService MultipleLines  \
count         7043   7043    7043       7043         7043          7043   
unique        7043      2       2          2            2             3   
top     3186-AJIEK   Male      No         No          Yes            No   
freq             1   3555    3641       4933         6361          3390   

       InternetService OnlineSecurity OnlineBackup DeviceProtection  \
count             7043           7043         7043             7043   
unique               3              3            3                3   
top        Fiber optic             No           No               No   
freq              3096           3498         3088             3095   

       TechSupport StreamingTV StreamingMovies        Contract  \
count         7043        7043            7043            7043   
unique           3           3               3               3   
top             No          No              No  Month-to-month   
freq          3473        2810            2785            3875   

       PaperlessBilling     PaymentMethod TotalCharges Churn  
count              7043              7043         7043  7043  
unique                2                 4         6531     2  
top                 Yes  Electronic check         20.2    No  
freq               4171              2365           11  5174

**Вывод:** Количество уникальных значение 7043 и каунт тоже, значит нет дубликотов, самый частый пол - мужской

## 3. Предобработка данных

### Дубликаты

Предобработайте датасет - проверьте на наличие дубликатов и удалите, если они есть.

In [ ]:
num_duplicates = df.duplicated().sum()
num_duplicates

np.int64(0)

### Отсутствующие значения

Пояснение к данному датасету:

Не всегда в данных есть явные пропуски, которые легко индексируются средствами Pandas. К сожалению бывают проблемы с выгрузками данных, некачественными источниками, или когда данные записываются вручную - часто присутствует человеческий фактов. По мере работы с новыми данными лучше проверять на наличие таких данных, иначе из-за этого исследование может исказиться.

Рассмотрим такой случай:

In [ ]:
df.isna().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

Да, не видно пропусков. Но почему тогда столбец `TotalCharges` не отобразился изначально как числовой столбец?

Проверим все столбцы:

In [ ]:
for col in df:
    if df[col].dtype == 'object':
        nans = df[col].apply(lambda x: len(x.strip())==0).sum()
        if nans > 0:
            print(f'Неявные пропуски столбца {col}:', df[col].apply(lambda x: len(x.strip())==0).sum())

Неявные пропуски столбца TotalCharges: 11


Что мы можем сделать в таком случае?

1) По этому же фильтру удалить строки с пропусками, чтобы не искажать данные
2) Перевести тип столбца на числовой и с параметром `errors = 'coerce'` и удалить

Но можно не удалять, чтобы потерять данные, а:

1) Заменить значения на среднюю
2) Поменять на другое - но только в том случае, если есть какая-то зависимость в данных.

Давайте изменим тип и найдем зависимость:

In [ ]:
df["TotalCharges"] = pd.to_numeric(df["TotalCharges"], errors = 'coerce')

df[df["TotalCharges"].isna()]

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
488   4472-LVYGI  Female              0     Yes        Yes       0   
753   3115-CZMZD    Male              0      No        Yes       0   
936   5709-LVOEQ  Female              0     Yes        Yes       0   
1082  4367-NUYAO    Male              0     Yes        Yes       0   
1340  1371-DWPAZ  Female              0     Yes        Yes       0   
3331  7644-OMVMY    Male              0     Yes        Yes       0   
3826  3213-VVOLG    Male              0     Yes        Yes       0   
4380  2520-SGTTA  Female              0     Yes        Yes       0   
5218  2923-ARZLG    Male              0     Yes        Yes       0   
6670  4075-WKNIU  Female              0     Yes        Yes       0   
6754  2775-SEFEE    Male              0      No        Yes       0   

     PhoneService     MultipleLines InternetService       OnlineSecurity  \
488            No  No phone service             DSL                  Yes   
753           Yes                No              No  No internet service   
936           Yes                No             DSL                  Yes   
1082          Yes               Yes              No  No internet service   
1340           No  No phone service             DSL                  Yes   
3331          Yes                No              No  No internet service   
3826          Yes               Yes              No  No internet service   
4380          Yes                No              No  No internet service   
5218          Yes                No              No  No internet service   
6670          Yes               Yes             DSL                   No   
6754          Yes               Yes             DSL                  Yes   

             OnlineBackup     DeviceProtection          TechSupport  \
488                    No                  Yes                  Yes   
753   No internet service  No internet service  No internet service   
936                   Yes                  Yes                   No   
1082  No internet service  No internet service  No internet service   
1340                  Yes                  Yes                  Yes   
3331  No internet service  No internet service  No internet service   
3826  No internet service  No internet service  No internet service   
4380  No internet service  No internet service  No internet service   
5218  No internet service  No internet service  No internet service   
6670                  Yes                  Yes                  Yes   
6754                  Yes                   No                  Yes   

              StreamingTV      StreamingMovies  Contract PaperlessBilling  \
488                   Yes                   No  Two year              Yes   
753   No internet service  No internet service  Two year               No   
936                   Yes                  Yes  Two year               No   
1082  No internet service  No internet service  Two year               No   
1340                  Yes                   No  Two year               No   
3331  No internet service  No internet service  Two year               No   
3826  No internet service  No internet service  Two year               No   
4380  No internet service  No internet service  Two year               No   
5218  No internet service  No internet service  One year              Yes   
6670                  Yes                   No  Two year               No   
6754                   No                   No  Two year              Yes   

                  PaymentMethod  MonthlyCharges  TotalCharges Churn  
488   Bank transfer (automatic)           52.55           NaN    No  
753                Mailed check           20.25           NaN    No  
936                Mailed check           80.85           NaN    No  
1082               Mailed check           25.75           NaN    No  
1340    Credit card (automatic)           56.05           NaN    No  
3331               Mailed check           19.85       

**Но как выглядят наши данные обычно?**

In [ ]:
df.sample(5)

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
860   4795-UXVCJ    Male              0      No         No      26   
570   7293-LSCDV  Female              0     Yes        Yes      60   
4519  0103-CSITQ  Female              0     Yes         No      57   
1195  4468-KAZHE  Female              1     Yes         No      20   
5628  6821-BUXUX  Female              0      No         No      13   

     PhoneService MultipleLines InternetService       OnlineSecurity  \
860           Yes            No              No  No internet service   
570           Yes            No              No  No internet service   
4519          Yes           Yes     Fiber optic                   No   
1195          Yes            No             DSL                   No   
5628          Yes           Yes     Fiber optic                  Yes   

             OnlineBackup     DeviceProtection          TechSupport  \
860   No internet service  No internet service  No internet service   
570   No internet service  No internet service  No internet service   
4519                  Yes                  Yes                  Yes   
1195                  Yes                   No                   No   
5628                  Yes                   No                   No   

              StreamingTV      StreamingMovies        Contract  \
860   No internet service  No internet service        One year   
570   No internet service  No internet service        Two year   
4519                  Yes                  Yes        One year   
1195                   No                  Yes  Month-to-month   
5628                  Yes                   No  Month-to-month   

     PaperlessBilling              PaymentMethod  MonthlyCharges  \
860                No    Credit card (automatic)           19.80   
570               Yes    Credit card (automatic)           19.25   
4519              Yes  Bank transfer (automatic)          109.40   
1195              Yes           Electronic check           60.00   
5628              Yes           Electronic check           96.65   

      TotalCharges Churn  
860         457.30    No  
570        1103.25    No  
4519       6252.70    No  
1195       1259.35    No  
5628       1244.50   Yes

**Ничего не заметили?**

Может общая сумма, списанная с клиента, равна сумме, взимаемой с клиента ежемесячно умноженное на количество времени?

Давайте посмострим.

In [ ]:
df[(df['TotalCharges'] / df['MonthlyCharges']) == df['tenure']].head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0   7590-VHVEG  Female              0     Yes         No       1           No   
20  8779-QRDMV    Male              1      No         No       1           No   
22  1066-JKSGK    Male              0      No         No       1          Yes   
27  8665-UTDHZ    Male              0     Yes        Yes       1           No   
33  7310-EGVHZ    Male              0      No         No       1          Yes   

       MultipleLines InternetService       OnlineSecurity  \
0   No phone service             DSL                   No   
20  No phone service             DSL                   No   
22                No              No  No internet service   
27  No phone service             DSL                   No   
33                No              No  No internet service   

           OnlineBackup     DeviceProtection          TechSupport  \
0                   Yes                   No                   No   
20                   No                  Yes                   No   
22  No internet service  No internet service  No internet service   
27                  Yes                   No                   No   
33  No internet service  No internet service  No internet service   

            StreamingTV      StreamingMovies        Contract PaperlessBilling  \
0                    No                   No  Month-to-month              Yes   
20                   No                  Yes  Month-to-month              Yes   
22  No internet service  No internet service  Month-to-month               No   
27                   No                   No  Month-to-month               No   
33  No internet service  No internet service  Month-to-month               No   

                PaymentMethod  MonthlyCharges  TotalCharges Churn  
0            Electronic check           29.85         29.85    No  
20           Electronic check           39.65         39.65   Yes  
22               Mailed check           20.15         20.15   Yes  
27           Electronic check           30.20         30.20   Yes  
33  Bank transfer (automatic)           20.20         20.20    No

Нашей гипотезе есть место быть.

Но давайте проверим:

In [ ]:
print('Количество совпадений по нашей теории:')
print(df[(df['TotalCharges'] / df['MonthlyCharges']) == df['tenure']]['customerID'].count())
print('Количество различий:')
print(df[(df['TotalCharges'] / df['MonthlyCharges']) != df['tenure']]['customerID'].count())

Количество совпадений по нашей теории:
614
Количество различий:
6429


**Как видно**, что различий очень много, относительно одинаковых значений.

Но, а если мы проверим наше различие в долях или процентах? Проверяем:

In [ ]:
# cоздаем series и считаем TotalCharges вручную
new_total = df['MonthlyCharges'] * df['tenure']

# посчитаем во сколько раз в среднем отличается наш new_total и TotalCharges - в процентном отношении
perc = abs(100 - (new_total / df['TotalCharges']).mean() * 100)

print(f'В среднем new_total отличается от TotalCharges на {perc:.4f}%')

В среднем new_total отличается от TotalCharges на 0.2311%


Как видно, они различаются незначительно (меньше `1%`), от чего мы можем предположить, что `TotalCharges` = `MonthlyCharges` * `время`. Скорее всего такие маленькие различия связаны с тем, что `tenure` округлялся.

В данном датасете указано только количество месяцев, без уточнения времени (до дней), потому будет грубо просто перемножить один столбец на другой. Но пропуски мы уже можем заменить на `0`, так как в тех строках `tenure` == 0. Такое небольшое исследование было сделано для того, чтобы заполнить наши пропуски нужным числом.

Заменим пропуски на нужное нам число.

In [ ]:
df['TotalCharges'].fillna(0, inplace=True)

C:\Users\korsh624.VIRO\AppData\Local\Temp\ipykernel_17216\171940771.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['TotalCharges'].fillna(0, inplace=True)


### Изменение типа данных

Рассмотрев все данные, замените типы на нужные (при необходимости):

- Если есть числа - на `int` или `float`
- Если категории - можно оставить `object`

In [ ]:
count=0
for i in range(df.shape[1]):
    try:
        df[i]=pd.to_numeric(df[i])
    except Exception as e:
        count+=1
        print(e)
count
df.info()


    

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


**Вывод:**

# 4 Проверка гипотез

Задание: сравнить сравнить две выборки - `Churn='No'` и `Churn='Yes'` 

Гипотеза - среднее количество месяцев, в течение которых клиент оставался в компании, отличается между двумя выборками.

In [6]:
group_no_churn = df[df['Churn'] == 'No']['tenure']
group_yes_churn = df[df['Churn'] == 'Yes']['tenure']
t_stat, p_value = stats.ttest_ind(group_no_churn, group_yes_churn)
print(f'Значение t-статистики: {t_stat}')
print(f'P-value: {p_value}')
# Интерпретация результатов
alpha = 0.05
if p_value < alpha:
    print('Существенные статистические различия между группами')
else:
    print('Средние значения не различаются статистически значимо')

Значение t-статистики: 31.57955051135377
P-value: 7.99905796059022e-205
Существенные статистические различия между группами


**Вывод:**Существенные статистические различия между группами

## 5 Статистический анализ на норму

Задание: выберите числовые признаки и проведите анализ - соответсвуют ли распредения нормальному?

In [ ]:
numeric_features=['SeniorCitizen','tenure','MonthlyCharges','TotalCharges']
for feature in numeric_features:
    # Проведение теста Шапиро-Уилка
    stat, p_value = stats.shapiro(df[feature])
    
    # Оценка результатов теста
    alpha = 0.05
    if p_value > alpha:
        print(f'Для признака {feature} p-value={p_value:.4f} > {alpha}. Распределение можно считать нормальным.')
    else:
        print(f'Для признака {feature} p-value={p_value:.4f} <= {alpha}. Распределение не является нормальным.')

Для признака SeniorCitizen p-value=0.0000 <= 0.05. Распределение не является нормальным.
Для признака tenure p-value=0.0000 <= 0.05. Распределение не является нормальным.
Для признака MonthlyCharges p-value=0.0000 <= 0.05. Распределение не является нормальным.
Для признака TotalCharges p-value=0.0000 <= 0.05. Распределение не является нормальным.


c:\Users\korsh624.VIRO\AppData\Local\Programs\Python\Python312\Lib\site-packages\scipy\stats\_axis_nan_policy.py:573: UserWarning: scipy.stats.shapiro: For N > 5000, computed p-value may not be accurate. Current N is 7043.
  res = hypotest_fun_out(*samples, **kwds)


**Вывод:**Для признака SeniorCitizen p-value=0.0000 <= 0.05. Распределение не является нормальным.
Для признака tenure p-value=0.0000 <= 0.05. Распределение не является нормальным.
Для признака MonthlyCharges p-value=0.0000 <= 0.05. Распределение не является нормальным.
Для признака TotalCharges p-value=0.0000 <= 0.05. Распределение не является нормальным.